### ISSUE TITLE
Misleading exception when trying to use KIWI on yaml files without the anymarkup python module

### ISSUE THREAD

**Issue by davidcassany** [OSInside/kiwi#1702]
## Problem description

KIWI trying to load yaml files without having the anymarkup python module installed on the host produces a misleading exception and traces.

```
kiwi-ng image info --description kdesc/suse/x86_64/suse-tumbleweed-docker/ --print-xml
[ INFO    ]: 17:43:48 | Reading runtime config file: /etc/kiwi.yml
[ INFO    ]: 17:43:48 | Reading runtime config file: /etc/kiwi.yml
[ INFO    ]: 17:43:48 | Loading XML description
[ INFO    ]: 17:43:48 | Support for XML markup available
[ ERROR   ]: 17:43:48 | Unexpected error:
Traceback (most recent call last):
  File "/usr/bin/kiwi-ng", line 11, in <module>
    load_entry_point('kiwi==9.21.23', 'console_scripts', 'kiwi-ng')()
  File "/usr/lib/python3.6/site-packages/kiwi/kiwi.py", line 68, in main
    App()
  File "/usr/lib/python3.6/site-packages/kiwi/app.py", line 35, in __init__
    task_class().process()
  File "/usr/lib/python3.6/site-packages/kiwi/tasks/image_info.py", line 73, in process
    self.command_args['--description']
  File "/usr/lib/python3.6/site-packages/kiwi/tasks/base.py", line 149, in load_xml_description
    config_file
  File "/usr/lib/python3.6/site-packages/kiwi/xml_description.py", line 70, in __init__
    self.description = self.markup.get_xml_description()
  File "/usr/lib/python3.6/site-packages/kiwi/markup/xml.py", line 37, in get_xml_description
    self.description
  File "/usr/lib/python3.6/site-packages/kiwi/markup/base.py", line 73, in apply_xslt_stylesheets
    etree.tostring(xslt_transform(etree.parse(description)))
  File "src/lxml/etree.pyx", line 3469, in lxml.etree.parse
  File "src/lxml/parser.pxi", line 1839, in lxml.etree._parseDocument
  File "src/lxml/parser.pxi", line 1865, in lxml.etree._parseDocumentFromURL
  File "src/lxml/parser.pxi", line 1769, in lxml.etree._parseDocFromFile
  File "src/lxml/parser.pxi", line 1163, in lxml.etree._BaseParser._parseDocFromFile
  File "src/lxml/parser.pxi", line 601, in lxml.etree._ParserContext._handleParseResultDoc
  File "src/lxml/parser.pxi", line 711, in lxml.etree._handleParseResult
  File "src/lxml/parser.pxi", line 640, in lxml.etree._raiseParseError
  File "kdesc/suse/x86_64/suse-tumbleweed-docker/config.kiwi", line 1
lxml.etree.XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1
```

KIWI, if anymarkup module is not available, tries to load the description as an XML file. This is good for backward compatibility, but probably we should throw some sort of warning that anymarkup is not there and that it will try to load an XML regardless of the file content.

## Expected behaviour

Get a human readable message about missing `anymarkup` module or any other sort of message that points into the right direction.

## Steps to reproduce the behaviour

Just run 
```
kiwi-ng image info --description kiwi-descriptions/suse/x86_64/suse-tumbleweed-docker/ --print-xml
```
using the descriptions from [OSInside/kiwi-descriptions](https://github.com/OSInside/kiwi-descriptions)

## OS and Software information
+ KIWI version: 9.21.23 (also happening on master)
+ Operating system: Leap 
+ OBS version:

**Comment by schaefi** [comment]
I agree the trace is not nice, so this should be better handled

However the error message

```
lxml.etree.XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1
```

Is the result of providing non XML data as input. Without anymarkup it would be not so easy to come up with a better message. We can catch "lxml.etree.XMLSyntaxError" and assume it's not XML and provide an error message. However if you provide XML and it has just a markup error the error message we provide would then again be misleading

So this is not easy and I would say we it's ok to catch the SyntaxError but don't try to be clever on it

Thoughts ?

**Comment by davidcassany** [comment]
The idea I had, for which I am not very enthusiastic, is to just throw some warning if `anymarkup` is not available making explicit that the description will be load as it was an XML and that no other formats will be understood. This way, in case the user is trying to load a yaml, they would know that the `anymarkup` is missing and that the description was loaded as an XML.

But yes, this does not avoid the confusing trace.

> So this is not easy and I would it's ok to catch the SyntaxError but don't try to be clever on it

Sure, trying to be clever here will be worse. I am more in the direction to try to give a human readable possible explanation of what could be the origin of such a trace.

In addition I am also wondering if shouldn't we switch to XML the docker description from [OSInside/kiwi-descriptions](https://github.com/OSInside/kiwi-descriptions/blob/master/suse/x86_64/suse-tumbleweed-docker/config.kiwi) or add some sort of comment within the description that it requires anymarkup python module. I like having a yaml example, however I believe it is likely to fail for user approaching KIWI and also hard to understand why it is failing.

**Comment by davidcassany** [comment]
Another idea could be to only try to guess the format with the descriptions files that are named `*.kiwi` and stick to XML for `*.xml` or YAML for `*.yaml` (or `*.yml`). I am also not very enthusiastic of this option and this implies a behavior change about how description files are being loaded...

**Comment by schaefi** [comment]
>The idea I had, for which I am not very enthusiastic, is to just throw some warning if anymarkup is not available making explicit that the description will be load as it was an XML and that no other formats will be understood. This way, in case the user is trying to load a yaml, they would know that the anymarkup is missing and that the description was loaded as an XML.

This sounds good to me and can be a warning. Be aware that this warning will show up probably most of the time because anymarkup is an optional thing and we don't force it

>Another idea could be to only try to guess the format with the descriptions files that are named *.kiwi and stick to XML for *.xml or YAML for *.yaml (or *.yml). I am also not very enthusiastic of this option and this implies a behavior change about how description files are being loaded...

Hmm, that would be a change in behavior and I think just for the freedom of a better error message I don't think we should do it

**Comment by davidcassany** [comment]
> Be aware that this warning will show up probably most of the time because anymarkup is an optional thing and we don't force it

Yes, that's why I am not very enthusiastic about it. It must be minimalist and not very disturbing. I'd say this should be sufficient if, in addition, we add some comments or hints to our yaml example. All this raised because there was someone trying KIWI to build docker images. This guy was using the yaml example and he could not figure out what was happening. As I am listed as a maintainer of image he contacted me.

**Comment by rtamalin** [comment]
I'll take this one.

Note that my first idea of using the Python magic module doesn't seem to be viable as the standard file utility just reports "ASCII Text" as the type for YAML files when I have tried it out on some Ansible and misc other YAML files ;-)

So this may not be so easy to do after all... Looking into alternatives that may be available.

**Comment by rtamalin** [comment]
So an interesting fact... The YAML safe_load() parser successfully loads an XML file as a string blob. This means that YAML's safe_load() succeeding will not differentiate between XML and non-XML... However the magic mechanism will successfully detect if a file is an XML document... So I should be able to make something work that way. ;-)

**Comment by rtamalin** [comment]
@schaefi @davidcassany So checking all the `*.xml` files on my Leap 15.2 box with the `file` command I get the following:
```% file -L $(locate -e -i -r "[.]xml$") | tr -s ' ' | grep -v -e "No such file or directory" | cut -d' ' -f 2- | sort | uniq -c
    277 ASCII text
      3 ASCII text, with no line terminators
      7 ASCII text, with very long lines
      8 empty
    164 exported SGML document, ASCII text
      2 exported SGML document, ASCII text, with very long lines
     10 exported SGML document, UTF-8 Unicode text
    716 HTML document, ASCII text
      7 HTML document, ASCII text, with very long lines
     48 HTML document, UTF-8 Unicode text
      1 HTML document, UTF-8 Unicode text, with very long lines
     16 UTF-8 Unicode text
      2 UTF-8 Unicode text, with very long lines
   2703 XML 1.0 document, ASCII text
     29 XML 1.0 document, ASCII text, with CRLF line terminators
     68 XML 1.0 document, ASCII text, with very long lines
      3 XML 1.0 document, ISO-8859 text
      7 XML 1.0 document, Non-ISO extended-ASCII text
     63 XML 1.0 document text
   1429 XML 1.0 document, UTF-8 Unicode text
    149 XML 1.0 document, UTF-8 Unicode text, with very long lines
      1 XML document, ASCII text
      1 XML document, Big-endian UTF-16 Unicode text, with CRLF line terminators
      1 XML document, UTF-8 Unicode (with BOM) text
```

Looking at some of the ones for which the file type is just reported as `ASCII Text` or `UTF-8 Unicode text` they appear to be HTML/XML blobs.

Therefore relying on file magic to determine if the content is safe to parse with the XML markup parser is probably not fully reliable...

So, in summary, I think that just reporting the warning that due to anymarkup not being available, the provided file will be parsed as XML, and if that isn't correct then the extension should be changed to a more appropriate value.

**Comment by rtamalin** [comment]
I'll put up a patch to implement that suggestion later. ;-)

**Comment by schaefi** [comment]
>Therefore relying on file magic to determine if the content is safe to parse with the XML markup parser is probably not fully reliable...

looks like. Interesting that there are so many variants. 

>So, in summary, I think that just reporting the warning that due to anymarkup not being available, the provided file will be parsed >as XML, and if that isn't correct then the extension should be changed to a more appropriate value.

ok, regarding the extension I should mention the type of files kiwi looks for. There is

* config.xml
* _some-file_name_.kiwi

kiwi only searches for these two variants. When customers use another than the XML markup the common approach is
often to have a link setup ```foo_bar.yaml -> config.kiwi```. But this is only a convention. So basically if anymarkup is not there and the given file is not named config.xml and fails to render as XML some good error message is appreciated :)

**Comment by schaefi** [comment]
sorry the draft came in while I was writing the message :) Thanks for the PR, we are looking at it asap

**Comment by rtamalin** [comment]
Hehe no worries... So I think what I have here improves things... But there may be room for further improvement...

With this change I get ```(kiwi_broken) fifteen[fergal] OSinside/kiwi % kiwi-ng image info --description /tmp/blah --print-xml
[ INFO    ]: 11:12:53 | Reading runtime config file: '/etc/kiwi.yml'
[ INFO    ]: 11:12:53 | Loading XML description
[ WARNING ]: 11:12:53 | Support for non-XML formatted config files requires the Python anymarkup module.
[ INFO    ]: 11:12:53 | Processing '/tmp/blah/config.kiwi' as XML
[ ERROR   ]: 11:12:53 | Unexpected error:
Traceback (most recent call last):
  File "/home/fergal/.venvs/kiwi_broken/bin/kiwi-ng", line 33, in <module>
    sys.exit(load_entry_point('kiwi', 'console_scripts', 'kiwi-ng')())
  File "/home/fergal/work/github/OSinside/kiwi/kiwi/kiwi.py", line 68, in main
    App()
  File "/home/fergal/work/github/OSinside/kiwi/kiwi/app.py", line 35, in __init__
    task_class().process()
  File "/home/fergal/work/github/OSinside/kiwi/kiwi/tasks/image_info.py", line 76, in process
    self.command_args['--description']
  File "/home/fergal/work/github/OSinside/kiwi/kiwi/tasks/base.py", line 150, in load_xml_description
    config_file
  File "/home/fergal/work/github/OSinside/kiwi/kiwi/xml_description.py", line 70, in __init__
    self.description = self.markup.get_xml_description()
  File "/home/fergal/work/github/OSinside/kiwi/kiwi/markup/xml.py", line 37, in get_xml_description
    self.description
  File "/home/fergal/work/github/OSinside/kiwi/kiwi/markup/base.py", line 58, in apply_xslt_stylesheets
    etree.tostring(xslt_transform(etree.parse(description)))
  File "src/lxml/etree.pyx", line 3521, in lxml.etree.parse
  File "src/lxml/parser.pxi", line 1859, in lxml.etree._parseDocument
  File "src/lxml/parser.pxi", line 1885, in lxml.etree._parseDocumentFromURL
  File "src/lxml/parser.pxi", line 1789, in lxml.etree._parseDocFromFile
  File "src/lxml/parser.pxi", line 1177, in lxml.etree._BaseParser._parseDocFromFile
  File "src/lxml/parser.pxi", line 615, in lxml.etree._ParserContext._handleParseResultDoc
  File "src/lxml/parser.pxi", line 725, in lxml.etree._handleParseResult
  File "src/lxml/parser.pxi", line 654, in lxml.etree._raiseParseError
  File "/tmp/blah/config.kiwi", line 1
lxml.etree.XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1
``` when testing with a venv with anymarkup removed...

It still has the ugly exception, but the warning and log message help explain it better.

**Comment by rtamalin** [comment]
So if this looks good, is this sufficient? Or do I need to add some form of pytest? I ask because I'm not sure how best to go about doing that so would welcome pointers to any relevant examples.

**Comment by schaefi** [comment]
Fixed with commit: #e57b57b90e641533c5f264af461be1428bb341eb
